### PR TITLE
[clang-format] Fix a bug in annotating TrailingReturnArrow

### LINKED ISSUE #69234: Clang-Format misformats Thread Safety Analysis annotations that include an arrow
For the following snippet:

```cpp
#if __has_attribute(capability)
#define LOCKABLE(name) __attribute__((capability(name)))
#endif

#if __has_attribute(unlock_function)
#define UNLOCK_FUNCTION(...) __attribute__((unlock_function(__VA_ARGS__)))
#endif

struct LOCKABLE("M") M {
  void unlock() UNLOCK_FUNCTION();
};

class L {
public:
  void f() UNLOCK_FUNCTION(this->mutex_) { this->mutex_.unlock(); }

private:
  M mutex_;
};
```

#67780 changed the behavior where it would insert extra spaces around the arrow in the annotation, so the following:
```cpp
  void f() UNLOCK_FUNCTION(this->mutex_) { this->mutex_.unlock(); }
```
becomes:
```cpp
  void f() UNLOCK_FUNCTION(this -> mutex_) { this->mutex_.unlock(); }
```

### PR DESCRIPTION
**Pull request by owenca**
Skip TrailingAnnotation when looking for TrailingReturnArrow.

Fixes #69234.

### CODE REVIEW

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -3497,6 +3497,15 @@ void TokenAnnotator::calculateFormattingInformation(AnnotatedLine &Line) const {
           Tok->setType(TT_TrailingReturnArrow);
           break;
         }
+        if (Tok->isNot(TT_TrailingAnnotation))
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:3500]: Why do you skip here?

File: clang/lib/Format/TokenAnnotator.cpp [hunk 2]
```diff
@@ -3497,6 +3497,15 @@ void TokenAnnotator::calculateFormattingInformation(AnnotatedLine &Line) const {
...
+          continue;
+        const auto *Next = Tok->Next;
+        if (!Next || Next->isNot(tok::l_paren))
+          continue;
+        Tok = Next->MatchingParen;
+        if (Tok)
+          continue;
+        break;
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:3508]: I find it rather confusing to have a loop, with a `break` at the end.
```suggestion
        if (!Tok)
          break;
```

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:3500]: Because `TT_TrailingAnnotation` is only the known case that needs special handling.

**Comment by owenca** [comment]
@HazardyKnusperkeks any other concerns?

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (HazardyKnusperkeks) [reply clang/lib/Format/TokenAnnotator.cpp:3500]: But why? If there is no `Next` or it isn't a `l_paren` we also `continue`, shouldn't all that match to `TT_TrailingAnnotation`?

**Review by HazardyKnusperkeks — APPROVED** [review]

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:3500]: Before: look for _any_ arrow after the `r_paren` of the function declaration.
Now: look for any arrow _except_ when inside the parentheses after a `TT_TrailingAnnotation`.

The case where `Next` is `nullptr` is handled by the condition of the `for` loop whereas any tokens (including non-`l_parens`) before a `TT_TrailingAnnotation` should be skipped as before.

**Comment by mydeveloperday** [comment]
Phyllis

**Review by mydeveloperday — APPROVED** [review]

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (HazardyKnusperkeks) [reply clang/lib/Format/TokenAnnotator.cpp:3500]: Understood.

**Pull request merged**